### ISSUE TITLE
Rare deadlock in `GraphNode.GetOrComputeValue`

### ISSUE THREAD

**Issue by majocha** [dotnet/fsharp#18115]
`GraphNode.GetOrComputeValue` can deadlock when the threadpool is under some pressure.

See: https://dev.azure.com/dnceng-public/public/_build/results?buildId=886415&view=logs&j=966a6946-91f0-5397-81f4-d998467880f9&t=344839a9-08c6-5829-65fb-26f32c5a016f&l=17626

**Repro steps**

This is reproducible in #17872 locally. For example:
in VS select all `BuildGraphTests` in Test Explorer and "run until failure".
The tests run in parallel and after few dozens or some more iterations the test case `Many requests to get a value asynchronously might evaluate the computation more than once even when some requests get canceled` will get stuck here
https://github.com/dotnet/fsharp/blob/a90f6d1b7573053e385a008da589233397bec48b/tests/FSharp.Compiler.Service.Tests/BuildGraphTests.fs#L221

**Comment by T-Gro** [comment]
Thank you Jakub for finding this.
Is it right to assume you only spotted this with parallel tests?

**Comment by majocha** [comment]
> Thank you Jakub for finding this. Is it right to assume you only spotted this with parallel tests?

Yes, it showed up once in CI recently. I confirmed a repro locally but also only in a parallel test run.

**Comment by majocha** [comment]
The issue is subtle. We attach a continuation that sets the taken flag:
https://github.com/dotnet/fsharp/blob/2c3bdcb3b6b9038acd76ecae5f4721614e71ea23/src/Compiler/Facilities/BuildGraph.fs#L57-L65

`ExecuteSynchronously` is there to ensure it is always executed before we get to this finally block:

https://github.com/dotnet/fsharp/blob/2c3bdcb3b6b9038acd76ecae5f4721614e71ea23/src/Compiler/Facilities/BuildGraph.fs#L88-L90 
*But* `SemaphoreSlim` [enforces asynchronous continuations](https://github.com/dotnet/runtime/commit/c5a6c24ee1b339872458c9fd5a93ed4af4c47893), overriding `ExecuteSynchronously`.

In effect, on rare occasions when the cancellation coincides with taking the samaphore, the finally block is executed before setting the flag. As a result, the semaphore is never released.
